### DISCUSSION TITLE
magit-am: Apply buffer containing a patch created by git-format-patch

### DISCUSSION THREAD

**Discussion by nbarrientos** [magit/magit#4906] [Ideas]
Hi,

Sometimes I have to review and incorporate patches sent to me using `git-send-email`. What I normally do is to open the mail which contains them using mu4e, then switch to the magit status buffer of the repository I want to apply the patches to and finally call this command:

```emacs-lisp
  (defun my/magit-am-apply-from-mu4e-article (&optional args)
    "Call `magit-am-apply-maildir' using the email in buffer *Article* as input."
    (interactive (list (magit-am-arguments)))
    (if-let* ((path (with-current-buffer "*Article*"
                      (mu4e-message-field-at-point :path))))
        (magit-am-apply-maildir path args)
      (user-error "No *Article* buffer found")))
```

which [I have bound](https://git.sr.ht/~nbarrientos/dotfiles/tree/87befb94099264640d066b18450527d8ba2ac255/.emacs.d/init.el#L1180) in the `magit-am` transient. This is clearly coupled to mu4e and how it displays e-mail but it does the job for me.

I was wondering though if you would consider a patch to achieve the same goal but with a different approach. What do you think about adding another action to the `magit-am` transient to apply (as mbox/maildir) an arbitrary buffer? Something like:

```emacs-lisp
(defun magit-am-apply-buffer (buffer &optional args)
  "Apply the patch in BUFFER."
  (interactive (list (read-buffer "Apply contents of buffer: ")
                     (magit-am-arguments)))
  (let ((temp-file
         (make-temp-file
          "magit."
          nil
          nil
          (with-current-buffer buffer
            (buffer-substring-no-properties (point-min) (point-max))))))
    (magit-am-apply-maildir temp-file args)))
```
This is a more flexible approach that could work with other mail user agents. Just point Magit to the buffer displaying the e-mail containing a patch formatted and sent using `git-send-email` and voila, it's applied. It leaves indeed a temporary file behind ;)

Please let me know, happy to send a pull request.

**Comment by kyleam** [comment]
There's some related discussion in gh-4099, which was about feeding a buffer `git apply` rather than `git am`.  It's just one opinion (and Jonas's is the one that really counts here), but [here's](https://github.com/magit/magit/pull/4099#issuecomment-619484403) what I said there.

---

As a side remark/pointer to another approach, I'd be happy if a mu4e user contributed an integration library to [piem](https://docs.kyleam.com/piem/), if that looks interesting to you.  (I've sadly left that project sit dormant for too long, but I do intend to start moving it along again.)

**Comment by nbarrientos** [comment]
Thanks for your input, @kyleam. Sorry, I wasn't aware a similar thing had been attempted before. Your comment there makes sense. I understand it's hard to have a good balance between how useful Magit is and where the boundaries between Magit and other Emacs bits/packages are set. Indeed if this goes forward, why not having another option to apply a "simple patch" from a buffer?

I'd be happy if we left things as there are now, I already have something that works (pretty tailored to what I need so quite efficient -- `w r`, patch applied, job done). I just wanted to propose to integrate into Magit a more generic approach (my proposal is not tied to any package) so more people could benefit from it, after all applying patches sent out with `git-send-email` sounds to me like very common use-case. I wouldn't surprised if somebody else knocked on the same door in the future.

I'll take a look at your pointer whenever I have a bit of free time.

Thanks again.

**Comment by nbarrientos** [comment]
And a side note for myself, just realised that `git-am` reads from stdin so indeed the temporary file could be avoided.

```emacs-lisp
(defun magit-am-apply-buffer (buffer &optional args)
  "Apply the patch in BUFFER."
  (interactive (list (read-buffer "Apply contents of buffer: ")
                     (magit-am-arguments)))
  (let ((patch
         (with-current-buffer buffer
           (buffer-substring-no-properties (point-min) (point-max)))))
    (with-temp-buffer
      (insert patch)
      (magit-run-git-with-input
       "am" args "-"))
    (magit-refresh)))
```

**Comment by nbarrientos** [comment]
I'll keep this functionality in my local configuration, thanks anyway.
